### PR TITLE
Add support for per-column defaults

### PR DESCRIPTION
**Pull request by funkybob**
Column.default defaults to missing, so everything will work as before by default.

If a default is passed, it's used on init.
If it's callable, it's called each time it's used - good for timestamps and UUIDs.

### CODE REVIEW

**Comment by numberoverzero** [comment]
First, thanks for the PR!  I'm sorry my response has been so slow.  That isn't due to a lack of interest in adding the feature; defaults are useful and if Bloop doesn't provide them, almost everyone will end up writing their own `Column` or `BaseModel` wrapper to expose them.


While this is a good implementation, I propose the following slight alternative.  This avoids putting more behavior in the `__init__` method of the base model, which retains the ability to use `BaseModel` as a mixin to an existing class, as well as supporting users with a different instantation point for bloop via `BaseModel.Meta.init`.  This will also help a bit with the still-in-planning abstract models #72.


Finally, this shouldn't establish precident for adding other flags or constraints to the `Column` class.  For example, the `Column` class doesn't need a `nullable` param that throws when loading/saving a model with a missing value.  This is better left to libraries built for those purposes, like marshmallow.

---

Let me know what you think of the following; if you'd prefer, I can code this up and add tests/docs.

## Usage

This is the same as the PR proposes; default is defined and stored on the Column object.

```python
from bloop import BaseModel, Column, Integer, String

NEXT_ID = 0

def monotonic_counter():
    global NEXT_ID
    id = NEXT_ID + 1
    NEXT_ID += 2
    return id


class User(BaseModel):
    id = Column(Integer, hash_key=True, default=monotonic_integer)
    name = Column(String, name="n")
    nickname = Column(String, name="k")
    age = Column(Integer, default=0)
```


## Implementation

This keeps the default value on the `Column` object, but moves the fallback logic into `Column.get` which is called on `Column.__get__` instead of occuring during instantiation.


```python
class Column:
    def __init__(self, ..., default=missing):
        ...
        self.default = default

    def get(self, obj):
        try:
            return super().get(obj)
        except AttributeError:
            default = self.default
            if default is not missing:
                if callable(default):
                    default = default()
                setattr(obj, self.name, default)
                return default
            raise
```

**Comment by funkybob** [comment]
Sure, moving it to __get__ / get makes sense... LGTM

**Comment by numberoverzero** [comment]
After a little implementation, I think there are three edge cases I missed initially.  Consider the following model, with the `default` implementation above:

```python
class User(BaseModel):
    id = Column(Integer, hash_key=True, default=monotonic_counter)
    verified = Column(Boolean, default=False)
    email = Column(String)
    secret = Column(String, default=random_token_func)
```

### When are defaults evaluated?

1. Should defaults be recorded on first use, or first save?  For example, assuming `monotonic_counter` always increments by 1 and we've not called it yet, the following would happen with no additional work than my proposal above:

```
>>> user = User()
>>> engine.save(user)

# invoked AFTER saving to DynamoDB
>>> user.secret
'qnluza'

>>> same_user = User(id=user.id)
>>> engine.load(same_user)

>>> same_user.secret
'oinwqf'
```

### Missing due to query projection

2. What happens when we query or scan and the projection doesn't include a column which provides a default?  Because the column wasn't part of the projected attributes, we don't know if a lack of value means it doesn't exist, or simply wasn't returned.  Consider a query against the following GSI for the model above:

```python
    by_email = GlobalSecondaryIndex(projection=[verified], hash_key=email)
```

Now a query against `User.by_email` won't include the user's secret, and a new one will be generated.

### Deleted values

What happens when I delete `user.secret` above?  Should the Column call the function again and provide a new default value, or should I get an `AttributeError` because it's been deleted?

---

I can solve the first one easily by changing the update rendering code to get marked columns and default columns ([`get_marked`](https://github.com/numberoverzero/bloop/blob/933fa9f102c5f53b9c5a1979d99bb73c18a8aef2/bloop/conditions.py#L365) becomes [`get_marked_and_default`](https://gist.github.com/numberoverzero/e67b767ace0571d4fd7aa044e0ad314c#file-update_with_defaults-py)) but the latter is a bit trickier.  There we'd want to know that the object has been loaded from DynamoDB at least once, and the column was never marked, so a default shouldn't be applied.  For the third, I'll need an explicit sentinel for deleted values similar to the existing `Sentinel("missing")` which would be used to suppress the default-loading behavior if the column was deleted.

Skipping the third issue for now, I think the logic is then ([rough code here](https://gist.github.com/numberoverzero/e67b767ace0571d4fd7aa044e0ad314c#file-update_with_defaults_sometimes_eager-py)):

1. Has the object ever been persisted to DynamoDB?
    * **NO**: This is a new object, so eagerly evaluate all defaults and persist them.
    * **YES**: This was loaded from or saved to DynamoDB, so only grab defaults from marked columns.
    

This simplifies if we only record defaults when they're used locally, but that could force awkward code where someone codes in `getattr(my_obj, "some_property")` for the side effect of storing a default, which is a code smell.

**Comment by xuru** [comment]
Well, I'll throw in my 2 cents...

I think what _*I'm*_ most use to, is the same model that [Marshmallow.fields](https://marshmallow.readthedocs.io/en/latest/api_reference.html#module-marshmallow.fields) uses.  That is, it only uses the default when saving, and the field is missing.  It shouldn't change the actual value of the field until it's actually saved.  That avoids any confusion as to why there is suddenly a value in a field that was never set, and allows for something like this:

```
class User(BaseModel):
    id = Column(UUID, hash_key=True, default=uuid.uuid4().hex)
    created = Column(DateTime, default=datetime.datetime.utcnow())
    modified = Column(DateTime, default=datetime.datetime.utcnow())
    active = Column(Boolean, default=True)

>>> user = User()
>>> same_user.id
None
>>> same_user.created
None
>>> engine.save(user)
>>> same_user = User(id=user.id)
>>> engine.load(same_user)
>>> same_user.id
'81a90422e12a414ab1264ff5c2cace48'
>>> same_user.created
2017-06-01 17:42:38.046997+00:00

```

**Comment by numberoverzero** [comment]
In your example, each instance will get the same default id and created/modified value.  Is that intended, or did you want each instance to have a different value?  Here's the version that passes each function instead:

```python
class User(BaseModel):
    id = Column(UUID, hash_key=True, default=uuid.uuid4)
    created = Column(DateTime, default=datetime.datetime.utcnow)
    modified = Column(DateTime, default=datetime.datetime.utcnow)
    active = Column(Boolean, default=True)
```

What happens when I delete a user's created time and save it again?  Do I get a new value, or is it saved without a created time?

```pycon
>>> user = User()
>>> user.created is None
True
>>> engine.save(user)
>>> user.created
2017-06-01 17:42:38.046997+00:00
>>> user.created = None
>>> user.created is None
True
>>> engine.save(user)
>>> user.created is None
# What happens here?  Still None because it was explicitly set?  Default function called again?
```

**Comment by xuru** [comment]
> In your example, each instance will get the same default id and created/modified value. Is that intended, or did you want each instance to have a different value?

Nope, yours is what I had in mind.  Thanks for the correction.

> What happens when I delete a user's created time and save it again? Do I get a new value, or is it saved without a created time?

Well, we can look at two examples...  Marshmallow and Flywheel.

## Marshmallow
For the Marshmallow example, it checks for a missing value at serialization time, and fills the value with whatever is default.  In our case, serialization would be saving to the DB.  So in your example above, after setting `created` to `None`, and calling `save()`, it would see the `None` as missing, apply the default, and save it with a new date.

## Flywheel
However, in the package Flywheel, it does it during creation time.  Here is the relevant code:
https://github.com/stevearc/flywheel/blob/master/flywheel/models.py#L177

So, in Flywheel, when the model is _created_ (whether instantiation or query/scan), it runs through the fields (and marks them as dirty so that it will be saved later), and applies defaults before returning the created object.

## Conclusion
I personally think the way that Flywheel does it is more expected when I set a default on a field, then Marshmallow's way of doing it IMHO

**Comment by numberoverzero** [comment]
tl;dr -- I'm not going to add defaults to the Column class, but I will add a section to the docs under Bloop Patterns that links to this PR and includes one example of implementing defaults.  Keep in mind that setting defaults in `__init__` **may cause data loss** if you query a projection which does not include the column with a default.

-----

At first glance defaults seem easy to add; evaluating them during `__init__` will work for most people most of the time.  Because it's such a common thing to do (booleans defaulting to `False`, datetimes to `now()`) I didn't expect much ambiguity around use.

Because the basic functionality is simple, there are multiple ways for defaults to work, and each feels "obvious" if that's the way your library or libraries have done things.  The central issue is mostly around how late-binding the evaluation is; for some cases like datetimes `now()` can vary significantly between `__init__` and `engine.save` and the difference is meaningful.  There are options that let users maintain control (lazy unless explicitly invoked) but this obscures the real problem:

If an implementation doesn't align with a user's expectation for how defaults work, the implementation will silently do the wrong thing and the user will have almost no way to know until it's too late.

Stepping away from local instantiation for a minute and talking about querying from DynamoDB, this is critically broken.  Take the following hypothetical model:

```python
class User(BaseModel):
    id = Column(String, hash_key=True)
    age = Column(Integer)
    email = Column(String)
    verified = Column(Boolean, default=False)

    by_email = GlobalSecondaryIndex(
        projection="keys", hash_key="email")
```

If I save an instance and then query on `by_email`, which only includes `id` in its projection, I won't get a value back for `verified`, **even if the row has a value for that column**:


```python
>>> user = User(id="mary", verified=True, email="mary@")
>>> engine.save(user)

>>> q = engine.query(User.by_email, key=User.email=="mary@")
>>> same_user = q.one()
```

At this point if I look at `same_user.verified` I can't know if the lack of value means the row in DynamoDB didn't have a value, or it simply wasn't loaded.  In this situation we could:

* set the default again, potentially losing the state from DynamoDB
* raise an error on seeing that the column wasn't loaded from DynamoDB
* not set the default, increasing the complexity of how defaults work

The first is unacceptable; Bloop will not guess or overwrite data. The second means that getting an attribute anywhere can now raise a new exception type that users need to handle any time they access an attribute, which is tedious and error prone.  The third means that our "obvious" implementation of defaults is no longer obvious, and has subtle cases where a column with a default may sometimes not have a value, and not call the default function to get one.

-----

If we evaluate how defaults interact with each engine operation, there are no cases where a single implementation will be obvious, safe, and consistent.  However, there is a strong argument for leaving defaults to be handled by the user:

1. `engine.save` after creating an object locally -- The local create can be wrapped in a function that applys defaults -- `user = User()` becomes `user = new_user()`.
2. `engine.save` after loading with `engine.load` -- called while updating an object that was loaded.  Defaults should already have been applied during the first save to DynamoDB, and applying them now would be application-specific logic and can't be dictated by bloop anyway.
3. `engine.query` on a model with a projection -- bloop can't know if a default is needed or not.  Providing one means potential data loss, and not providing one means Columns with defaults many not always have defaults.


-----

I've [put up a gist here](https://gist.github.com/numberoverzero/42606974b007fa0dda1646fd5094b59e) which includes a model and a controller, demonstrating how/when I set defaults.  If this seems useful let me know and I'll link it from the Patterns section.  I'll also take suggestions for a smaller snippet to include inline for eg. a `create_user` function, or even a generic `create_entity` function which takes defaults and applies them to `**kwargs` before passing to the constructor.

**Comment by numberoverzero** [comment]
Closing this; Bloop supports defaults with `Column(default=<value or function>)` in the `dev-2.0` branch.  Sorry for the long back and forth on this one; I'm glad we came to a consistent, simple behavior that doesn't interact poorly with projections.
